File: dfs/112_Path_Sum.cpp
/**
 * @file    112_Path_Sum.cpp
 *
 * @author  Juan Diego Becerra
 * @date    2021-11-01
 * @brief   https://leetcode.com/problems/path-sum/
 *
 */

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    bool hasPathSum(TreeNode *root, int targetSum)
    {
        if (root)
        {
            targetSum -= root->val;

            if (!root->left && !root->right)
            {
                return targetSum == 0;
            }
            return hasPathSum(root->left, targetSum) || hasPathSum(root->right, targetSum);
        }
        return false;
    }
};

File: dfs/112_Path_Sum.py
"""112_Path_Sum.py

Author: Juan Diego Becerra
Date:   06-30-2022
Brief:  https://leetcode.com/problems/path-sum/
"""


# Time: O(h) | Space: O(1)
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# Time: O(h) | Space: O(1)
def hasPathSum(self, root: TreeNode, targetSum: int) -> bool:
    if not root:
        return False

    if not (root.left or root.right):
        return targetSum - root.val == 0

    return hasPathSum(root.left, targetSum - root.val) or hasPathSum(root.right, targetSum - root.val)